File: ruby/labs/beginner/lab1-basics/solution.rb
# Lab 1: Ruby Basics & Object-Oriented Programming
# Complete Solution

class Book
  attr_accessor :title, :author, :isbn

  def initialize(title, author, isbn)
    @title = title
    @author = author
    @isbn = isbn
  end

  def to_s
    "\"#{@title}\" by #{@author} (ISBN: #{@isbn})"
  end

  def matches_title?(search_term)
    @title.downcase.include?(search_term.downcase)
  end

  def matches_author?(search_term)
    @author.downcase.include?(search_term.downcase)
  end
end

class Library
  attr_reader :name, :books

  @@all_libraries = []

  def initialize(name)
    @name = name
    @books = []
    @@all_libraries << self
  end

  def self.all
    @@all_libraries
  end

  def self.total_books
    @@all_libraries.sum { |library| library.book_count }
  end

  def add_book(book)
    @books << book
    puts "✓ Added: #{book}"
  end

  def book_count
    @books.length
  end

  def find_by_title(search_term)
    @books.select { |book| book.matches_title?(search_term) }
  end

  def find_by_author(search_term)
    @books.select { |book| book.matches_author?(search_term) }
  end

  def list_all_books
    if @books.empty?
      puts "The library is empty."
    else
      puts "\n#{@name} - Books:"
      @books.each_with_index do |book, index|
        puts "  #{index + 1}. #{book}"
      end
    end
  end

  def statistics
    puts "\n#{@name} Statistics:"
    puts "  Total books: #{book_count}"
    puts "  Unique authors: #{@books.map(&:author).uniq.length}"
  end
end

# Main program
puts "=" * 60
puts " " * 15 + "LIBRARY MANAGEMENT SYSTEM"
puts "=" * 60

# Create first library
lib1 = Library.new("City Central Library")
puts "\nCreated: #{lib1.name}"

# Add books to first library
lib1.add_book(Book.new("The Ruby Way", "Hal Fulton", "978-0321714633"))
lib1.add_book(Book.new("Eloquent Ruby", "Russ Olsen", "978-0321584106"))
lib1.add_book(Book.new("Programming Ruby", "Dave Thomas", "978-1934356081"))

# Create second library
lib2 = Library.new("University Library")
puts "\nCreated: #{lib2.name}"

# Add books to second library
lib2.add_book(Book.new("Metaprogramming Ruby", "Paolo Perrotta", "978-1941222126"))
lib2.add_book(Book.new("Ruby Under a Microscope", "Pat Shaughnessy", "978-1593275273"))
lib2.add_book(Book.new("Confident Ruby", "Avdi Grimm", "978-0692040201"))
lib2.add_book(Book.new("Design Patterns in Ruby", "Russ Olsen", "978-0321490452"))

# List all books in each library
puts "\n" + "=" * 60
puts " " * 20 + "ALL BOOKS"
puts "=" * 60
lib1.list_all_books
lib2.list_all_books

# Search for books
puts "\n" + "=" * 60
puts " " * 18 + "SEARCH RESULTS"
puts "=" * 60

puts "\nSearching for books with 'Ruby' in title at #{lib1.name}:"
results = lib1.find_by_title("ruby")
if results.empty?
  puts "  No books found."
else
  results.each { |book| puts "  → #{book}" }
end

puts "\nSearching for books by 'Olsen' at #{lib2.name}:"
results = lib2.find_by_author("olsen")
if results.empty?
  puts "  No books found."
else
  results.each { |book| puts "  → #{book}" }
end

# Display statistics
puts "\n" + "=" * 60
puts " " * 22 + "STATISTICS"
puts "=" * 60
lib1.statistics
lib2.statistics

puts "\n" + "-" * 60
puts "Total libraries: #{Library.all.length}"
puts "Total books across all libraries: #{Library.total_books}"
puts "=" * 60

File: dart/labs/beginner/lab1-basics/solution.dart
// Lab 1: Dart Basics & Object-Oriented Programming
// Complete Solution

class Book {
  String title;
  String author;
  String? isbn; // Nullable for flexibility

  // Named parameters for clarity
  Book({
    required this.title,
    required this.author,
    this.isbn,
  });

  @override
  String toString() {
    var isbnText = isbn != null ? '(ISBN: $isbn)' : '(No ISBN)';
    return '"$title" by $author $isbnText';
  }

  bool matchesTitle(String searchTerm) {
    return title.toLowerCase().contains(searchTerm.toLowerCase());
  }

  bool matchesAuthor(String searchTerm) {
    return author.toLowerCase().contains(searchTerm.toLowerCase());
  }
}

class Library {
  String name;
  List<Book> books;

  Library(this.name) : books = [];

  void addBook(Book book) {
    books.add(book);
    print('✓ Added: $book');
  }

  int get bookCount => books.length;

  int get uniqueAuthorsCount {
    var authors = books.map((book) => book.author).toSet();
    return authors.length;
  }

  List<Book> findByTitle(String searchTerm) {
    return books.where((book) => book.matchesTitle(searchTerm)).toList();
  }

  List<Book> findByAuthor(String searchTerm) {
    return books.where((book) => book.matchesAuthor(searchTerm)).toList();
  }

  void listAllBooks() {
    if (books.isEmpty) {
      print('The library is empty.');
    } else {
      print('\n$name - Books:');
      for (var i = 0; i < books.length; i++) {
        print('  ${i + 1}. ${books[i]}');
      }
    }
  }

  void displayStatistics() {
    print('\n$name Statistics:');
    print('  Total books: $bookCount');
    print('  Unique authors: $uniqueAuthorsCount');
  }
}

void main() {
  print('=' * 70);
  print(' ' * 20 + 'LIBRARY MANAGEMENT SYSTEM');
  print('=' * 70);

  // Create first library
  var lib1 = Library('City Central Library');
  print('\n📚 Created: ${lib1.name}');

  // Add books to first library
  lib1.addBook(Book(
    title: 'The Dart Way',
    author: 'Kathy Walrath',
    isbn: '978-0321927705',
  ));
  lib1.addBook(Book(
    title: 'Dart in Action',
    author: 'Chris Buckett',
    isbn: '978-1617290862',
  ));
  lib1.addBook(Book(
    title: 'Dart Up and Running',
    author: 'Kathy Walrath',
    isbn: '978-1449330897',
  ));

  // Create second library
  var lib2 = Library('University Library');
  print('\n📚 Created: ${lib2.name}');

  // Add books to second library
  lib2.addBook(Book(
    title: 'Flutter in Action',
    author: 'Eric Windmill',
    isbn: '978-1617296147',
  ));
  lib2.addBook(Book(
    title: 'Programming Dart',
    author: 'Chris Buckett',
    isbn: '978-1937785277',
  ));
  lib2.addBook(Book(
    title: 'Learning Dart',
    author: 'Dzenan Ridjanovic',
    isbn: '978-1849697415',
  ));
  lib2.addBook(Book(
    title: 'Dart for Absolute Beginners',
    author: 'David Kopec',
    isbn: '978-1430264811',
  ));

  // List all books in each library
  print('\n' + '=' * 70);
  print(' ' * 28 + 'ALL BOOKS');
  print('=' * 70);
  lib1.listAllBooks();
  lib2.listAllBooks();

  // Search for books
  print('\n' + '=' * 70);
  print(' ' * 26 + 'SEARCH RESULTS');
  print('=' * 70);

  print('\n🔍 Searching for books with "Dart" in title at ${lib1.name}:');
  var results = lib1.findByTitle('dart');
  if (results.isEmpty) {
    print('  No books found.');
  } else {
    for (var book in results) {
      print('  → $book');
    }
  }

  print('\n🔍 Searching for books by "Buckett" at ${lib2.name}:');
  results = lib2.findByAuthor('buckett');
  if (results.isEmpty) {
    print('  No books found.');
  } else {
    for (var book in results) {
      print('  → $book');
    }
  }

  // Display statistics
  print('\n' + '=' * 70);
  print(' ' * 30 + 'STATISTICS');
  print('=' * 70);
  lib1.displayStatistics();
  lib2.displayStatistics();

  print('\n' + '-' * 70);
  print('Total libraries: 2');
  print('Total books across all libraries: ${lib1.bookCount + lib2.bookCount}');
  print('=' * 70);
}
